submissions/VBystrov/tiny-world-oop/index.js

```diff
@@ -0,0 +1,110 @@
+/* Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details
+   Complete the below for code reviewers' convenience:
+
+   Code repository: https://github.com/VBystrov/a-tiny-JS-world
+   Web app: https://vbystrov.github.io/a-tiny-JS-world/
+   */
+
+// ======== OBJECTS DEFINITIONS ========
+// Define your objects here
+
+class Creature {
+  constructor(name, gender = 'unknown') {
+    this.name = name;
+    this.gender = gender;
+    this.friends = [];
+  }
+
+  addFriends(...newFriends) {
+    newFriends.forEach(({ name }) => this.friends.push(name));
+    return this.friends;
+  }
+}
+
+class Dog extends Creature {
+  constructor(name, gender) {
+    super(name, gender);
+    this.species = 'dog';
+    this.legs = 4;
+    this.saying = 'woof!';
+  }
+}
+
+class Cat extends Creature {
+  constructor(name, gender) {
+    super(name, gender);
+    this.species = 'cat';
+    this.legs = 4;
+    this.saying = 'mew-mew-mew-mew';
+  }
+}
+
+class Human extends Creature {
+  constructor(name, gender) {
+    super(name, gender);
+    this.species = 'homo sapiens';
+    this.legs = 2;
+    this.hands = 2;
+  }
+}
+
+class Woman extends Human {
+  constructor(name) {
+    super(name, 'female');
+    this.saying = 'I am a woman.';
+  }
+}
+
+class Man extends Human {
+  constructor(name) {
+    super(name, 'male');
+    this.saying = 'I am a man.';
+  }
+}
+
+const dog = new Dog('Globgor', 'male');
+const cat = new Cat('Eclipsa', 'female');
+const woman = new Woman('Star');
+const man = new Man('Marko');
+man.addFriends(cat, woman);
+woman.addFriends(man, dog);
+dog.addFriends(cat, woman, man);
+
+const creatures = [dog, cat, woman, man];
+
+// ======== OUTPUT ========
+/* Use print(message) for output.
+   Default tag for message is <pre>. Use print(message,'div') to change containing element tag.
+
+   Message can contain HTML markup. You may also tweak index.html and/or styles.css.
+   However, please, REFRAIN from improving visuals at least until your code is reviewed
+   so code reviewers might focus on a single file that is index.js.
+   */
+
+const getTemplate = (creature) => {
+  let template = '';
+  for (let key in creature) {
+    if (key === 'friends' && !creature[key].length) {
+      continue;
+    }
+    if (creature[key]) {
+      template += `${key}: ${creature[key]}; `;
+    }
+  }
+
+  return template;
+};
```

Comment:
Each class can represent itself.
Make this function a method of the base class and child classes.
Remember every class should refer to the properties it owns directly.
In a child class you can overload an inherited method and also benefit from calling an inherited method by means of `super.methodName()`. This way child's methods gets whatever is needed from a parent class and supplements it with its own data.

If method is named `toString` then this method is called implicitly in string context. I.e. `print(object.toString())` is equivalent of `print(object)` as `print` function ensures string context for its arguments.

Reply:
Done.